relay-threading/src/multiplexing.rs

```diff
@@ -0,0 +1,380 @@
+            // stream has ended, in that case we return.
+            let Poll::Ready(Some(result)) = this.futures.as_mut().poll_next(cx) else {
+                return;
+            };
+
+            // If there is an error, it means that the future has panicked, we want to notify this.
+            match (this.panic_handler.as_ref(), result) {
+                // Panic handler and error, we swallow the panic and invoke the callback.
+                (Some(panic_handler), Err(error)) => {
+                    panic_handler(error);
+                }
+                // No panic handler and error, we propagate the panic.
+                (None, Err(error)) => {
+                    std::panic::resume_unwind(error);
+                }
+                // Otherwise, we do nothing.
+                (_, Ok(())) => {}
+            }
+        }
+    }
+}
+
+pin_project! {
+    /// [`Multiplexed`] is a future that concurrently schedules asynchronous tasks from a stream while ensuring that
+    /// the number of concurrently executing tasks does not exceed a specified limit.
+    ///
+    /// This multiplexer is primarily used by the [`AsyncPool`] to manage task execution on worker threads.
+    ///
+    /// # Panics
+    ///
+    /// If any task panics without a custom panic handler, the panic will propagate.
+    pub struct Multiplexed<S, F> {
+        max_concurrency: usize,
+        #[pin]
+        rx: S,
+        #[pin]
+        tasks: Tasks<F>,
+    }
+}
+
+impl<S, F> Multiplexed<S, F>
+where
+    S: Stream<Item = F>,
+{
+    /// Creates a new [`Multiplexed`] instance with a defined concurrency limit and a stream of tasks.
+    ///
+    /// Tasks from the stream will be scheduled for execution concurrently, and an optional panic handler
+    /// can be provided to manage errors during task execution.
+    #[allow(clippy::type_complexity)]
+    pub fn new(
+        max_concurrency: usize,
+        rx: S,
+        panic_handler: Option<Arc<dyn Fn(Box<dyn Any + Send>) + Send + Sync>>,
+    ) -> Self {
+        Self {
+            max_concurrency,
+            rx,
+            tasks: Tasks::new(panic_handler),
+        }
+    }
+}
+
+impl<S, F> Future for Multiplexed<S, F>
+where
+    S: FusedStream<Item = F>,
+    F: Future<Output = ()>,
+{
+    type Output = ();
+
+    /// Polls the [`Multiplexed`] future to drive task execution.
+    ///
+    /// This method repeatedly schedules new tasks from the stream while enforcing the concurrency limit.
+    /// It completes when the stream is exhausted and no active tasks remain.
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut this = self.project();
+
+        loop {
+            this.tasks.as_mut().poll_tasks_until_pending(cx);
+
+            if this.tasks.len() >= *this.max_concurrency {
+                return Poll::Pending;
+            }
+
+            match this.rx.as_mut().poll_next(cx) {
```

Comment:
Are we allowed to poll the stream again when it already yielded `Poll::Ready(None)`?

Reply:
Yeah that's a good point, we should check whether the stream is terminated, but keep iterating until tasks are exhausted.